### PR TITLE
Zero-dimensional arrays FAQ

### PR DESCRIPTION
**Pull request by cossio**
Adding some documentation about zero-dimensional arrays. Some discussion on Slack:

https://julialang.slack.com/archives/C6GHSTN4T/p1529875426000022

I copied/pasted some comments by more knowledgeable people, with minor edits.

Questions about 0-dimensional arrays arise often. And I think they are a bit unintuitive. Some documentation might help.

https://discourse.julialang.org/t/what-is-the-reason-for-a-zero-dimensional-array/7175

https://discourse.julialang.org/t/why-have-0-dimensional-arrays/797

Hopefully this clarifies the concept for others as well.

Supersedes https://github.com/JuliaLang/julia/pull/27765.

### CODE REVIEW

**Review by StefanKarpinski — APPROVED** [review]
Very nice write up!

**Comment by cossio** [comment]
Thanks to @mbauman , @nalimilan , @KristofferC

File: doc/src/manual/faq.md [hunk 1]
```diff
@@ -732,6 +732,27 @@ julia> @sync for i in 1:3
 1 Foo  Bar 2 Foo  Bar 3 Foo  Bar
 ```
 
+## Arrays
+
+### Zero-dimensional arrays
+
+Zero-dimensional arrays are arrays of the form `Array{T,0}`. They behave similar to scalars, but there are important differences. They deserve a special mention because they are a special case which makes logical sense given the generic definition of arrays, but might be a bit unintuitive at first. The following line defines a zero-dimensional array:
```
Comment (nalimilan) [inline doc/src/manual/faq.md]: Break lines at 92 chars.

Comment (cossio) [reply doc/src/manual/faq.md]: I don't understand. Do I need to do something?

**Review by cossio — COMMENTED** [review]
Break long lines

File: doc/src/manual/faq.md (hunk 1 above)
Comment (cossio) [reply doc/src/manual/faq.md]: Done

**Review by cossio — COMMENTED** [review]
Done

**Comment by StefanKarpinski** [comment]
There are white space issues.

**Comment by cossio** [comment]
What exactly? What do I have to do?

**Comment by StefanKarpinski** [comment]
Run `make check-whitespace` and it will report that certain lines have trailing whitespace. Get rid of the trailing whitespace, get `make check-whitespace` passing and then push the fixed version.

**Comment by mbauman** [comment]
He's editing via GitHub, though, which does indeed make this difficult to see.  It's not ideal, but we have this rule since many contributors' editors do automatically remove whitespace in any file they touch… leading to unrelated changes, conflicts, and churn.

File: doc/src/manual/faq.md [hunk 2]
```diff
@@ -732,6 +732,47 @@ julia> @sync for i in 1:3
 1 Foo  Bar 2 Foo  Bar 3 Foo  Bar
 ```
 
+## Arrays
+
+### Zero-dimensional arrays
```
Comment (mbauman) [inline doc/src/manual/faq.md]: Now that this lives in a FAQ, let's make this a question.

**Comment by cossio** [comment]
Yep, I was editing in Github directly. Anyway I think I addressed the issues. Let me know if something else is needed.

**Review by mbauman — APPROVED** [review]

**Comment by mbauman** [comment]
Well, ideally Refs would also implement all the methods to make them behave largely like 0-dimensional arrays themselves (#26988).  But, I think you're right — typically folks don't need or run into them, which is precisely why they cause confusion when it does occur, and which is why it's nice to have it as a FAQ instead of a part of the narrative manual.  They do mostly just "fall out" of an implementation in a consistent manner; a good example is `view(A, 1)` which provides a view into the first element of an array.  Returning an unwrapped scalar would no longer be a view. I'd wager they're most likely obtained by accidentally forgetting args in `zeros()` or `Array{Int}(undef)` or some such.  

We could perhaps emphasize further the fact that 0-dimensional arrays are first and foremost _containers_ and do not behave like the element they wrap.  Furthermore, I had been also thinking about calling out the fact that we do not allow 0-dimensional arrays to participate in the linear algebra of matrices.  Thus, while you can broadcast `rand(3,3) .* fill(2)`, we have not defined the multiplication `rand(3,3) * fill(2)`.

Just checking in here: how you doing, @cossio?  Getting revision fatigue yet?  Let us know before you get too frustrated with us. I know incremental reviews can be a challenge.  :)

**Comment by cossio** [comment]
Doing fine @mbauman. Unfortunately I don't have the necessary knowledge to contribute much to this discussion.

**Comment by StefanKarpinski** [comment]
I think this is good as is. If someone wants to tweak the language to make zero-dimensional arrays sound less appealing after this is merged, they can do that. The new text addresses a common point of confusion well, so let's just merge it.

**Comment by cossio** [comment]
Why CI fails? I don't understand the error message:

` !! 'Arrays' is not unique. [src/manual/mathematical-operations.md]`

**Comment by fredrikekre** [comment]
I believe there is a clash between the header you are adding and some existing `## Arrays` header in the manual. Lets wait for #27769 to resolve that.

File: doc/src/manual/faq.md [hunk 3]
```diff
@@ -732,6 +732,50 @@ julia> @sync for i in 1:3
...
+* Zero-dimensional arrays are the "point" to vector's "line" and matrix's
+"plane". Just as a line has no area (but still represents a set of things), a
+point has no length or any dimensions at all (but still represents a thing).
+* We define `prod(())` to be 1, and the total number of elements in an array is
+the product of the size. The size of a zero-dimensional array is `()`, and
+therefore its length is `1`.
+* Zero-dimensional arrays don't natively have any dimensions into which you
+index -- they’re just `A[]`. We can apply the same "trailing one" rule for them
+as for all other array dimensionalities, so you can indeed index them as
+`A[1]`, `A[1,1]`, etc.
```
Comment (DrKrar) [inline doc/src/manual/faq.md]: I think  `A[1,1]` would produce an error

Comment (StefanKarpinski) [reply doc/src/manual/faq.md]: It doesn't:
```jl
julia> A = fill(1)
0-dimensional Array{Int64,0}:
1

julia> A[1,1]
1
```
Same rule that allows indexing a vector like an `n×1` matrix (or `n×1×1` three-tensor):
```jl
julia> v = rand(3)
3-element Array{Float64,1}:
 0.7636187688715259
 0.6218914012928478
 0.22322554192055066

julia> v[3,1]
0.22322554192055066
```

**Comment by nalimilan** [comment]
Locally the build passes with that PR after rebasing on master (now that #27769 is merged).

**Comment by cossio** [comment]
@nalimilan Should I do something?

**Comment by simonbyrne** [comment]
Thanks!

**Comment by cossio** [comment]
awesome!

**Comment by cossio** [comment]
@StefanKarpinski The bulleted list appears broken: https://docs.julialang.org/en/latest/manual/faq/#Arrays-1. I guess to fix that I would have to remove the line breaks in the source markdown, but that's agains the max-chars per line rule. How do I fix this?

**Comment by cossio** [comment]
@KristofferC  Thanks!

File: doc/src/manual/faq.md [hunk 4]
```diff
@@ -732,6 +732,50 @@ julia> @sync for i in 1:3
...
+
+It is also important to understand the differences to ordinary scalars. Scalars
+are not mutable containers (even though they are iterable and define things
```
Comment (andyferris) [inline doc/src/manual/faq.md]: Suggestion: "...(even though *some scalar types such as `Number`* are iterable..."

(I'm pretty sure broadcasting, `map`/`reduce`/`filter` of arrays of scalars, etc doesn't rely on this being defined for all scalar types?)

**Pull request merged**